### ISSUE TITLE
Un-needed (wrong?) code in example script

### ISSUE THREAD

**Issue by vittorio-perera** [huggingface/accelerate#956]
I am referring to [this](https://github.com/huggingface/accelerate/blob/f7427c86ee47d88490f09f8e5476592073f656a4/examples/by_feature/deepspeed_with_config_support.py#L288) line of code in [`deepspeed_with_config_support.py`](https://github.com/huggingface/accelerate/blob/main/examples/by_feature/deepspeed_with_config_support.py):
```
    ...
    losses.append(accelerator.gather_for_metrics(loss.repeat(args.per_device_eval_batch_size)))

losses = torch.cat(losses)
    try:
        eval_loss = torch.mean(losses)
        ...
```

In this case the loss is repeated by the batch size, gathered across processes and then averaged. While it is unclear to me why one would repeat the loss (perhaps to compute the average loss per sample?) the computation does not contribute to the final result (e.g., we repeat `[2, 4]` into `[2, 2, 4, 4]` just to get the final result of `3` anyway).

**Comment by pacman100** [comment]
Hello @vittorio-perera, it is needed when the final batch is uneven.

For example, losses were [2, 4] but the batch size on GPU 1 is 2 and the remaining final batch size on GPU 2 is 1. Then, `accelerator.gather_for_metrics` would result in [2, 2, 4] leading the mean to be 2.66 instead of 3.

**Comment by vittorio-perera** [comment]
@pacman100 thanks for the answer! There's something I still don't understand, maybe you can help me figure it out?  Let's say that I have a dataset with 7 samples, I am training with a batch size of 2 and I am using 2 GPUs. In this case we are in the same situation you described, where in the last batch we have 2 samples on GPU1 and a 1 sample on GPU2. Let's say that, as in your example, the losses for the two batches are 2 (for the batch on GPU1) and 4 (for the batch on GPU2). What should be the correct result 3 or 2.66? In other words, are we computing the average loss per batch or the average loss per sample?

Going back to the example script, the loss, for each gpu, is a scalar. Accordingly `accelerator.gather_for_metrics(loss)` will return a 2x1 tensor (e.g., `[2, 4]`) but you suggest it is a 3x1 tensor instead. The only way I was able to get a 3x1 tensor is when calling `accelerator.gather_for_metrics(loss.repeat(2))` (where 2 is the batch size I am considering). My understanding here is that two things happens:

1. `loss.repeat` creates two 2x1 tensors (i.e. `[2, 2]` and `[4, 4]`)
2. `accelerator.gather_for_metrics` reduces `[4, 4]` to `[4]` (as it's coming from a batch that wasn't full size) and then puts it together with `[2, 2]` to return `[2, 2, 4]` a 3x1 tensor. 

Can you confirm if this is what is happening?

**Comment by pacman100** [comment]
Hello @vittorio-perera , the script is computing average loss per sample and your understanding is correct.

**Comment by vittorio-perera** [comment]
Perfect this makes sense now! Thanks for bearing with me ☺️
